third_party/move/move-vm/integration-tests/src/tests/binary_format_version.rs

```diff
@@ -24,62 +27,98 @@ fn test_publish_module_with_custom_max_binary_format_version() {
             AccountAddress::from_hex_literal("0x1").unwrap(),
             move_stdlib::natives::GasParameters::zeros(),
-        ));
+        );
+        let runtime_environment = RuntimeEnvironment::new(natives.clone());
+        let vm = MoveVM::new(natives);
         let mut sess = vm.new_session(&storage);
 
-        sess.publish_module(
-            b_new.clone(),
-            *m.self_id().address(),
-            &mut UnmeteredGasMeter,
-        )
-        .unwrap();
+        if vm.vm_config().use_loader_v2 {
+            let module_storage = storage.as_unsync_module_storage(&runtime_environment);
+            let new_module_storage =
+                StagingModuleStorage::create(m.self_addr(), &module_storage, vec![b_new
+                    .clone()
+                    .into()])
+                .expect("New module should be publishable");
+            StagingModuleStorage::create(m.self_addr(), &new_module_storage, vec![b_old
+                .clone()
+                .into()])
+            .expect("Old module should be publishable");
+        } else {
+            #[allow(deprecated)]
+            sess.publish_module_bundle(
+                vec![b_new.clone()],
+                *m.self_id().address(),
+                &mut UnmeteredGasMeter,
+            )
+            .unwrap();
 
-        sess.publish_module(
-            b_old.clone(),
-            *m.self_id().address(),
-            &mut UnmeteredGasMeter,
-        )
-        .unwrap();
+            #[allow(deprecated)]
+            sess.publish_module_bundle(
+                vec![b_old.clone()],
+                *m.self_id().address(),
+                &mut UnmeteredGasMeter,
+            )
+            .unwrap();
+        }
     }
 
     // Should reject the module with newer version with max binary format version being set to VERSION_MAX - 1
     {
         let storage = InMemoryStorage::new();
-        let vm = MoveVM::new_with_config(
-            move_stdlib::natives::all_natives(
-                AccountAddress::from_hex_literal("0x1").unwrap(),
-                move_stdlib::natives::GasParameters::zeros(),
-            ),
-            VMConfig {
-                deserializer_config: DeserializerConfig::new(
-                    VERSION_MAX.checked_sub(1).unwrap(),
-                    IDENTIFIER_SIZE_MAX,
-                ),
-                ..Default::default()
-            },
+        let natives = move_stdlib::natives::all_natives(
+            AccountAddress::from_hex_literal("0x1").unwrap(),
+            move_stdlib::natives::GasParameters::zeros(),
         );
+        let vm_config = VMConfig {
+            deserializer_config: DeserializerConfig::new(
+                VERSION_MAX.checked_sub(1).unwrap(),
+                IDENTIFIER_SIZE_MAX,
+            ),
+            ..Default::default()
+        };
+        let runtime_environment =
+            RuntimeEnvironment::new_with_config(natives.clone(), vm_config.clone());
```

Comment:
Seems that RuntimeEnvironment and MoveVM both has natives and configs? Why duplication over there?

Reply:
Because we need to maintain both V1 and V2. V1 needs to also store natives, V2 needs only VM config. To avoid that, V2 stores an empty vec there.